source/beamline/CLS/CLSSIS3820ScalerModeControl.cpp

```diff
@@ -0,0 +1,270 @@
+	}
+
+	return result;
+}
+
+bool CLSSIS3820ScalerModeControl::validValue(double value) const
+{
+	bool result = false;
+
+	int valueInt = int(value);
+
+	if (valueInt == Continuous || valueInt == SingleShot || valueInt == None)
+		result = true;
+
+	return result;
+}
+
+bool CLSSIS3820ScalerModeControl::validSetpoint(double value) const
+{
+	bool result = false;
+
+	int valueInt = int(value);
+
+	if (valueInt == Continuous || valueInt == SingleShot)
+		result = true;
+
+	return result;
+}
+
+QString CLSSIS3820ScalerModeControl::toString() const
+{
+	QString basicInfo = AMPseudoMotorControl::toString();
+
+	// Note this control's current value.
+
+	QString controlValue = "Value: " + toString(value());
+
+	// Note this control's tolerance.
+
+	QString controlTolerance = "Tolerance: " + QString::number(tolerance());
+
+	// Create and return complete info string.
+
+	QString result = basicInfo + "\n" + controlValue + "\n" + controlTolerance;
+
+	return result;
+}
+
+QString CLSSIS3820ScalerModeControl::toString(double value) const
+{
+	QString result = "";
+
+	switch (int(value)) {
+	case SingleShot:
+		result = "Single Shot";
+		break;
+	case Continuous:
+		result = "Continuous";
+		break;
+	case None:
+		result = "None";
+		break;
+	default:
+		result = "Unknown";
+		break;
+	}
+
+	return result;
+}
+
+void CLSSIS3820ScalerModeControl::setScanCountControl(AMControl *newControl)
+{
+	if (scanCountControl_ != newControl) {
+
+		if (scanCountControl_)
+			removeChildControl(scanCountControl_);
+
+		scanCountControl_ = newControl;
+
+		if (scanCountControl_)
+			addChildControl(scanCountControl_);
+
+		emit scanCountControlChanged(scanCountControl_);
+	}
+}
+
+void CLSSIS3820ScalerModeControl::setNumberOfScansPerBufferControl(AMControl *newControl)
+{
+	if (numberOfScansPerBufferControl_ != newControl) {
+
+		if (numberOfScansPerBufferControl_)
+			removeChildControl(numberOfScansPerBufferControl_);
+
+		numberOfScansPerBufferControl_ = newControl;
+
+		if (numberOfScansPerBufferControl_)
+			addChildControl(numberOfScansPerBufferControl_);
+
+		emit numberOfScansPerBufferControlChanged(numberOfScansPerBufferControl_);
+	}
+}
+
+void CLSSIS3820ScalerModeControl::setStartScanControl(AMControl *newControl)
+{
+	if (startScanControl_ != newControl) {
+
+		if (startScanControl_)
+			removeChildControl(startScanControl_);
+
+		startScanControl_ = newControl;
+
+		if (startScanControl_)
+			addChildControl(startScanControl_);
+
+		emit startScanControlChanged(startScanControl_);
+	}
+}
+
+void CLSSIS3820ScalerModeControl::updateConnected()
+{
+	bool wasConnected = connected_;
+
+	bool isConnected = (
+				scanCountControl_ && scanCountControl_->isConnected() &&
+				numberOfScansPerBufferControl_ && numberOfScansPerBufferControl_->isConnected() &&
+				startScanControl_ && startScanControl_->isConnected()
+				);
+
+	setConnected(isConnected);
+
+	// If we weren't connected before but we are now, set mode to Continuous.
+
+	if (!wasConnected && isConnected)
```

Comment:
I would say that they probably shouldn't be changed at all.  If it's connected, it should check all the PVs to determine which state it is in, rather than forcing a new state.  This would also deal with your comment below about multiple instances of AM.

In general, I would say that nothing should change state on startup.  It should read what it's states are and try its best to tell you where it is.